rosidl_cmake/cmake/rosidl_generate_interfaces.cmake

```diff
@@ -137,22 +137,50 @@ macro(rosidl_generate_interfaces target)
   # afterwards all remaining interface files are .idl files
   list(APPEND _idl_tuples ${_idl_adapter_tuples})
 
-  # to generate action interfaces, we need to depend on "action_msgs"
+  set(_service_msgs_added FALSE)
+  set(_action_msgs_added FALSE)
   foreach(_tuple ${_interface_tuples})
     string(REGEX REPLACE ".*:([^:]*)$" "\\1" _tuple_file "${_tuple}")
     get_filename_component(_parent_dir "${_tuple_file}" DIRECTORY)
-    if("${_parent_dir}" STREQUAL "action")
+
+    if("${_parent_dir}" STREQUAL "action" AND NOT _action_msgs_added)
+      if (_service_msgs_added)
+        find_package(service_msgs QUIET)
+        if(NOT ${service_msgs_FOUND})
+          message(FATAL_ERROR
+            "Unable to generate service interface for '${_tuple_file}'. "
+            "In order to generate service interfaces you must add a depend tag "
+            "for 'service_msgs' in your package.xml.")
+        endif()
+        ament_export_dependencies(service_msgs)
+        list_append_unique(_ARG_DEPENDENCIES "service_msgs")
+        set(_service_msgs_added TRUE)
+      endif()
       find_package(action_msgs QUIET)
       if(NOT ${action_msgs_FOUND})
         message(FATAL_ERROR
           "Unable to generate action interface for '${_tuple_file}'. "
           "In order to generate action interfaces you must add a depend tag "
           "for 'action_msgs' in your package.xml.")
       endif()
-      list_append_unique(_ARG_DEPENDENCIES "action_msgs")
       ament_export_dependencies(action_msgs)
+      list_append_unique(_ARG_DEPENDENCIES "action_msgs")
+      set(_action_msgs_added TRUE)
       break()
+
+    elseif("${_parent_dir}" STREQUAL "srv" AND NOT _service_msgs_added)
+      find_package(service_msgs QUIET)
+      if(NOT ${service_msgs_FOUND})
+        message(FATAL_ERROR
+          "Unable to generate service interface for '${_tuple_file}'. "
+          "In order to generate service interfaces you must add a depend tag "
+          "for 'service_msgs' in your package.xml.")
+      endif()
+      ament_export_dependencies(service_msgs)
+      list_append_unique(_ARG_DEPENDENCIES "service_msgs")
+      set(_service_msgs_added TRUE)
     endif()
+
```

Comment:
you could add here an `if()` that breaks the loop if both service_msgs and action_msgs were already FOUND

Reply:
[91ea5a9](https://github.com/ros2/rosidl/pull/700/commits/91ea5a9062cecf44e8d77fc730c06ea89e922593)